R/pack.R

```diff
@@ -139,6 +146,87 @@ unpack <- function(data, cols, names_sep = NULL, names_repair = "check_unique")
   reconstruct_tibble(data, out)
 }
 
+check_inner_inner_duplicate <- function(x, error_call = caller_env()) {
+  n <- length(x)
+
+  if (n == 0L || n == 1L) {
+    # Nothing to duplicate across
+    return(invisible())
+  }
+
+  outer <- names(x)
+  x <- unname(x)
+
+  inner <- map(x, names)
+  inner <- map(inner, unique)
+
+  outer <- vec_rep_each(outer, times = list_sizes(inner))
+  inner <- list_unchop(inner, ptype = character())
+
+  problems <- vec_duplicate_detect(inner)
+
+  if (!any(problems)) {
+    return(invisible())
+  }
+
+  outer <- vec_slice(outer, problems)
+  inner <- vec_slice(inner, problems)
+
+  split <- vec_split(outer, inner)
+
+  inners <- split$key
+  outers <- split$val
+
+  bullets <- map2_chr(inners, outers, function(inner, outer) {
+    cli::format_inline("{.code {inner}}, across {.code {outer}}.")
+  })
+  bullets <- set_names(bullets, "i")
+  bullets <- cli::format_bullets_raw(bullets)
+  bullets <- set_names(bullets, " ")
+
+  message <- c(
+    "Can't duplicate names across the modified columns.",
+    x = "These names are duplicated:",
+    bullets,
+    i = "Use `names_sep` to disambiguate using the column name.",
+    i = "Or use `names_repair` to specify a repair strategy."
+  )
+
+  cli::cli_abort(message, call = error_call)
+}
+
+check_outer_inner_duplicate <- function(x, outer, error_call = caller_env()) {
+  # Names of unpacked columns will disappear so aren't considered
+  outer <- setdiff(outer, names(x))
+
+  inner <- map(x, names)
+  inner <- map(inner, unique)
+
+  problems <- map(inner, function(x) intersect(x, outer))
+  problems <- vec_slice(problems, list_sizes(problems) != 0L)
+
+  if (length(problems) == 0L) {
+    return(invisible())
+  }
+
+  bullets <- map2_chr(problems, names(problems), function(inner, outer) {
+    cli::format_inline("{.code {inner}}, from {.code {outer}}.")
+  })
+  bullets <- set_names(bullets, "i")
+  bullets <- cli::format_bullets_raw(bullets)
+  bullets <- set_names(bullets, " ")
+
+  message <- c(
+    "Can't duplicate names between the modified columns and the original data.",
```

Comment:
Is this consistent with other tidyr functions? Should the inner names replace the outer names?

Reply:
I feel like that might be more surprising than useful? i.e. the inner/outer collision is the reason for the first error you see here https://tidyr.tidyverse.org/articles/rectangle.html#sharla-gelfands-discography, but I feel like knowing there was a collision is pretty useful. And having `id` replaced "automatically" by the inner one might be surprising, especially if you aren't exactly sure what the names of those inner columns are.

It seems like `pivot_longer/wider()` also error, and they don't use `unpack()`.

``` r
tibble(x = c(1, 2), a = c(3, 4), b = c(5, 6)) %>%
  pivot_longer(c(a, b), names_to = "x")
#> Error in `pivot_longer()`:
#> ! Names must be unique.
#> ✖ These names are duplicated:
#>   * "x" at locations 1 and 2.
#> ℹ Use argument `names_repair` to specify repair strategy.

#> Backtrace:
#>      ▆
#>   1. ├─tibble(x = c(1, 2), a = c(3, 4), b = c(5, 6)) %>% ...
#>   2. ├─tidyr::pivot_longer(., c(a, b), names_to = "x")
#>   3. └─tidyr:::pivot_longer.data.frame(., c(a, b), names_to = "x") at tidyr/R/pivot-long.R:153:2
#>   4.   └─tidyr::pivot_longer_spec(...) at tidyr/R/pivot-long.R:185:2
#>   5.     ├─tidyr:::wrap_error_names(...) at tidyr/R/pivot-long.R:328:2
#>   6.     │ └─base::withCallingHandlers(...) at tidyr/R/pivot.R:57:2
#>   7.     └─vctrs::vec_cbind(...) at tidyr/R/pivot-long.R:328:2
#>   8.       └─vctrs (local) `<fn>`()
#>   9.         └─vctrs:::validate_unique(names = names, arg = arg, call = call)
#>  10.           └─vctrs:::stop_names_must_be_unique(names, arg, call = call)
#>  11.             └─vctrs:::stop_names(...)
#>  12.               └─vctrs:::stop_vctrs(...)
#>  13.                 └─rlang::abort(message, class = c(class, "vctrs_error"), ..., call = vctrs_error_call(call))

tibble(x = 1:2, a = c("x", "y"), b = c(3, 4)) %>%
  pivot_wider(names_from = a, values_from = b)
#> Error in `pivot_wider()`:
#> ! Names must be unique.
#> ✖ These names are duplicated:
#>   * "x" at locations 1 and 2.
#> ℹ Use argument `names_repair` to specify repair strategy.
```

Things like `unnest()`, `unnest_wider()`, `separate_*()` all go through `unpack()` so those will be consistent.

Reply:
Ok, sounds good.